### PR TITLE
Add WandB logging and hyperparameter sweep support

### PR DESCRIPTION
**Pull request by podestplatz**
## Summary
- Add optional WandB logger integration for experiment tracking
- Implement hyperparameter sweep functionality with WandB
- Add early stopping and dynamic learning rate scheduling

### CODE REVIEW

File: examples/01-Introduction/sweep_example.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    """
+    # Initialize wandb project
+    wandb.login()
+    
+    # Create the sweep
+    sweep_id = wandb.sweep(sweep_config, project="neuralhydrology")
+    
+    print(f"Created sweep with ID: {sweep_id}")
+    print(f"To run sweep agents, use:")
+    print(f"wandb agent {sweep_id}")
```
[the remark this answers is not in the document]
Comment (podestplatz) [reply examples/01-Introduction/sweep_example.py:74]: It would know what to run, if we would pass the "programm" attribute with a pointer to the `nh_run.py` script in the sweep configuration. However, I personally like to have this approach with dedicated scripts. 
Also, these prints are removed now as they have been a left-over from the prototyping phase.

File: neuralhydrology/training/wandb_logger.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    import wandb
+except ImportError as e:
+    raise ImportError("WandB is not installed. Run pip install -e .[wandb]") from e
+
+from neuralhydrology.__about__ import __version__
+from neuralhydrology.utils.config import Config
+from neuralhydrology.utils.logging_utils import get_git_hash, save_git_diff
+
+
+class Logger(object):
```
[the remark this answers is not in the document]
Comment (podestplatz) [reply neuralhydrology/training/wandb_logger.py:18]: See comment below.

File: neuralhydrology/utils/config.py [hunk 3]
```diff
@@ -542,10 +542,22 @@ def log_n_figures(self) -> int:
         else:
             return self._cfg["log_n_figures"]
 
+    @property
+    def logger_type(self) -> str:
+        return self._cfg.get("logger_type", "tensorboard")
+
     @property
     def log_tensorboard(self) -> bool:
         return self._cfg.get("log_tensorboard", True)
```
[the remark this answers is not in the document]
Comment (podestplatz) [reply neuralhydrology/utils/config.py:552]: Done. I thought, that without this property, the config file would not be able to be deserialised leading to backward-compatibility issues. But that does not seem to be the case.

File: neuralhydrology/training/basetrainer.py [hunk 4]
```diff
@@ -19,7 +19,7 @@
 from neuralhydrology.evaluation.tester import BaseTester
 from neuralhydrology.modelzoo import get_model
 from neuralhydrology.training import get_loss_obj, get_optimizer, get_regularization_obj
-from neuralhydrology.training.logger import Logger
+from neuralhydrology.training import tensorboard_logger, wandb_logger
```
[the remark this answers is not in the document]
Comment (podestplatz) [reply neuralhydrology/training/basetrainer.py:22]: you're right. I changed that.

File: neuralhydrology/training/basetrainer.py [hunk 5]
```diff
@@ -180,9 +180,15 @@ def initialize_training(self):
...
-        if self.cfg.log_tensorboard:
-            self.experiment_logger.start_tb()
+        if self.cfg.logger_type == "wandb":
+            self.experiment_logger = wandb_logger.Logger(cfg=self.cfg)
+        elif self.cfg.logger_type == "tensorboard":
+            self.experiment_logger = tensorboard_logger.Logger(cfg=self.cfg)
+        else:
+            raise ValueError(f"Invalid logger_type '{self.cfg.logger_type}'. Must be either 'wandb' or 'tensorboard'.")
+
+        if self.cfg.log_metrics:
```
[the remark this answers is not in the document]
Comment (podestplatz) [reply neuralhydrology/training/basetrainer.py:190]: done.

File: neuralhydrology/nh_run.py [hunk 6]
```diff
@@ -147,7 +155,105 @@ def finetune(config_file: Path = None, gpu: int = None):
...
+    gpu : int, optional
+        GPU id to use. Will override config argument 'device'. A value smaller than zero indicates CPU.
+        Don't use this argument if you want to use the device as specified in the config file e.g. MPS.
+    sweep_id : str, optional
+        The sweep ID to connect to. If not provided, it will be read from the WANDB_SWEEP_ID 
+        environment variable.
+
+    """
+    import wandb
+    import os
```
[the remark this answers is not in the document]
Comment (podestplatz) [reply neuralhydrology/nh_run.py:179]: done!

File: neuralhydrology/training/logging/wandb_logger.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        figures : List[mpl.figure.Figure]
+            List of figures to save.
+        freq : str
+            Prediction frequency of the figures.
+        preamble : str, optional
+            Prefix to prepend to the figures' file names.
+        """
+        for idx, figure in enumerate(figures):
+            self.run.log({f"validation/timeseries_{freq}_{idx + 1}": wandb.Image(figure)}, step=self.update)
+            figure.savefig(Path(self._img_log_dir, preamble + f'_freq{freq}_epoch{self.epoch}_{idx + 1}'), dpi=300)
```
Comment (gauchm) [inline neuralhydrology/training/logging/wandb_logger.py:66]: please extract this call into the parent class.

File: neuralhydrology/training/basetrainer.py [hunk 8]
```diff
@@ -180,9 +178,18 @@ def initialize_training(self):
...
+            self.experiment_logger = tensorboard_logger.TensorboardLogger(cfg=self.cfg)
+        elif self.cfg.logger_type is None:
+            pass  # No logging is done if no logger type is specified
+        else:
+            raise ValueError(f"Invalid logger_type '{self.cfg.logger_type}'. Must be either 'wandb' or 'tensorboard'.")
+
+        if self.cfg.logger_type is not None:
+            self.experiment_logger.start_logger()
 
         if self.cfg.is_continue_training:
```
Comment (gauchm) [inline neuralhydrology/training/basetrainer.py:194]: the calls here will fail if logger_type is None (experiment_logger will be None). Same for many of the calls below. Prior to this PR we used to always create a Logger (and just not start it), so the fact that experiment_logger can now be None changes things. Having lots of is not None checks is not nice though, so I wonder if we should create something like a `NoOpLogger` that does nothing. Just a thought, maybe you have better ideas.

If we do that, we can probably also get rid of all the writer is not None/run is not None checks in the logger implementations.

File: neuralhydrology/nh_wandb_sweep.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+"""
+This script configures a sweep and creates it on WandB. 
+
+Use nh_run.py in sweep mode to add an agent to the sweep. An agent will then execute as long as there are configurations
+left to try in the sweep. 
+"""
+import wandb
+
+# This specifies the hyperparameters to sweep over
+sweep_config = {
```
Comment (gauchm) [inline neuralhydrology/nh_wandb_sweep.py:11]: I fear that people will read this as the "neuralhydorlogy endorsed HP tuning configuration" and not think properly about what they want to sweep. Can we turn this into an empty dict and add a comment that points to the tutorial?

File: examples/07-WandB/WandB.ipynb [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "# Install dependencies\n",
+    "Independent of how you've installed 'neuralhydrology`, just run\n",
+    "```bash\n",
+    "pip install wandb\n",
+    "```\n",
+    "in you terminal to install. `neuralhydrology` also defines a group of optional dependencies called \"wandb\". So if you prefer installing the WandB dependencies `neuralhydrology` specifies then either run\n",
```
Comment (gauchm) [inline examples/07-WandB/WandB.ipynb:44]:
```suggestion
    "in your terminal to install. `neuralhydrology` also defines a group of optional dependencies called \"wandb\". So if you prefer installing the WandB dependencies `neuralhydrology` specifies then either run\n",
```

File: examples/07-WandB/WandB.ipynb [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+   "id": "22417d132844fa57"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "# Logging to WandB\n",
+    "The `neuralhydrology` repository supports tensorboard logging out of the box and provides WandB as an alternative (the actual file implementing the logging functionality is `neuralhydrology/training/wandb_logger.py`).\n",
+    "Logging is controlled through the following properties:\n",
+    "- `logger_type` defines whether the metrics shall be logged to `\"tensorboard\"` or `\"wandb\"`. Per default it is set to `tensorboard` and if you want to disable logging altogether set it to `null`, `~` or leaving it blank.\n",
```
Comment (gauchm) [inline examples/07-WandB/WandB.ipynb:62]:
```suggestion
    "- `logger_type` defines whether the metrics shall be logged to `\"tensorboard\"` or `\"wandb\"`. By default it is set to `tensorboard` and if you want to disable logging altogether set it to `null`, `~` or leave it blank.\n",
```

File: examples/07-WandB/WandB.ipynb [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+   ],
+   "id": "bce4488c8858c487",
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "Using this sweep configuration, we call `wandb.sweep()` pass the config to it and specify the name of the project again in which\n",
```
Comment (gauchm) [inline examples/07-WandB/WandB.ipynb:162]:
```suggestion
    "Using this sweep configuration, we call `wandb.sweep()`, pass the config to it and specify the name of the project again in which\n",
```

**Review by gauchm — COMMENTED** [review]
Some minor comments, plus the tests are failing. I didn't have time to look into those but I think it's related to my comment about experiment_logger being None.

File: neuralhydrology/training/basetrainer.py (hunk 8 above)
Comment (podestplatz) [reply neuralhydrology/training/basetrainer.py:193]: Yeah, good point. As I did not come up with a better idea, I implemented the NoOpLogger and removed the is_running checks in both classes.
